### PR TITLE
ignore any number of leading underscores when converting from snake case to camel case

### PR DESCRIPTION
**Pull request by lsmith77**
ie. leave '_username' unchanged

fixes https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1009

### CODE REVIEW

File: Normalizer/CamelKeysNormalizer.php [hunk 1]
```diff
@@ -75,8 +75,17 @@ private function normalizeString($string)
             return $string;
         }
 
-        return preg_replace_callback('/_([a-zA-Z0-9])/', function ($matches) {
+        if (preg_match('/^(_+)(.*)/', $string, $matches)) {
```
Comment (lsmith77) [inline Normalizer/CamelKeysNormalizer.php]: or should we just ignore anything with a leading underscore? or should we just ignore a single underscore?

**Comment by Ener-Getick** [comment]
Why don't we use ``Serializer::deserialize`` ?

**Comment by Ener-Getick** [comment]
I'm not sure that we should keep this feature in 2.0 as it's not intuitive and I don't really see the interest of converting the request parameters from snake case to camel case...

**Comment by lsmith77** [comment]
reason for its existence is explained here https://florian.voutzinos.com/blog/handling-camelcase-with-fosrestbundle/

**Comment by Ener-Getick** [comment]
hum... What about creating a black list instead of changing the behaviour for all params ?

**Comment by lsmith77** [comment]
yeah .. I guess this can indeed break things .. I do not think that a black list is the way to go .. but we can either delay this until 1.8 or even 2.0 .. or we can provide a switch to get this new behavior.

**Comment by Ener-Getick** [comment]
Well I'm still not sure this is a good solution to have an array normalizer on top of the request parameters... As we can see, it can break many things.
IMO it would be better to at least specify on each request if we want to use the normalizer (via a request matcher or an annotation in the controller) or to create a new ``FormSubmitter`` providing a method to convert the parameters from snake case to camel case.

**Comment by lsmith77** [comment]
in 2.0 we have a way to enable listeners for specific paths

**Comment by Ener-Getick** [comment]
Yeah, so wouldn't it be better to orientate people to 2.0 instead of changing the array normalizer behavior?

**Comment by lsmith77** [comment]
well it fixes something I would consider a bug and it has been reported as a bug.
now we can say this bug was there so long it has become a "feature" that would be a BC break to fix. in this case we can offer it as a config option or we provide 2 services, ie. add a new service with the new behavior.

**Comment by Ener-Getick** [comment]
I personally would prefer a new service/option. 
What do you think @xabbuh?

**Comment by lsmith77** [comment]
moved it to a new class/service

File: Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+     *
+     * @return string
+     */
+    protected function normalizeString($string)
+    {
+        if (false === strpos($string, '_')) {
+            return $string;
+        }
+
+        if (preg_match('/^(_+)(.*)/', $string, $matches)) {
```
Comment (Ener-Getick) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: IMO you should add a ``$`` to be sure that you match the entire string

**Comment by Ener-Getick** [comment]
:+1:

Status: reviewed

File: Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php (hunk 2 above)
Comment (lsmith77) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: sure .. actually I am wondering if there is a more efficient way than to use a regexp here ..

Comment (Ener-Getick) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: Sure you can maybe use strspn http://php.net/manual/fr/function.strspn.php

Comment (Ener-Getick) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: You can also replace the preg_replace by an explode with an array_reduce

Comment (lsmith77) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: good call .. updated the code accordingly

Comment (lsmith77) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: yeah .. but I doubt that this will be more efficient or more readable ..

Comment (Ener-Getick) [inline Normalizer/CamelKeysNormalizerWithLeadingUnderscore.php]: We have to benchmark it ;-)

**Pull request merged**